ml/backend.go

```diff
@@ -26,19 +26,28 @@ type Backend interface {
 	SystemInfo() string
 }
 
-var backends = make(map[string]func(*os.File) (Backend, error))
+// BackendParams controls how the backend loads and executes models
+type BackendParams struct {
+	NumThreads int // Number of threads to use if running on the CPU
 
-func RegisterBackend(name string, f func(*os.File) (Backend, error)) {
+	MainGPU      int       // Main GPU
+	NumGPULayers int       // Number of layers to offload to GPU
+	TensorSplit  []float32 // Fraction of the model to offload to each GPU
+}
+
+var backends = make(map[string]func(*os.File, BackendParams) (Backend, error))
+
+func RegisterBackend(name string, f func(*os.File, BackendParams) (Backend, error)) {
 	if _, ok := backends[name]; ok {
 		panic("backend: backend already registered")
 	}
 
 	backends[name] = f
 }
 
-func NewBackend(f *os.File) (Backend, error) {
+func NewBackend(f *os.File, params BackendParams) (Backend, error) {
 	if backend, ok := backends["ggml"]; ok {
```

Comment:
Are any of the params required? Wondering if we should check anything here and fail fast.

Reply:
None of the ones defined here are required, if they are not set then the backend can have default behavior that it defines. For example, on Apple Silicon these parameters aren't really that useful with just a single GPU and unified memory.